python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -414,51 +426,137 @@ def execute_asset_backfill_iteration(
+            run_tags=backfill.tags,
+            backfill_start_time=backfill_start_time,
+        ):
+            yield None
+
+        if not isinstance(result, AssetBackfillIterationResult):
+            check.failed(
+                "Expected execute_asset_backfill_iteration_inner to return an"
+                " AssetBackfillIterationResult"
+            )
+
+        updated_backfill = backfill.with_asset_backfill_data(
+            result.backfill_data, dynamic_partitions_store=instance
         )
+        if result.backfill_data.is_complete():
+            # The asset backfill is complete when all runs to be requested have finished (success,
+            # failure, or cancellation). Since the AssetBackfillData object stores materialization states
+            # per asset partition, the daemon continues to update the backfill data until all runs have
+            # finished in order to display the final partition statuses in the UI.
+            updated_backfill = updated_backfill.with_status(BulkActionStatus.COMPLETED)
 
-    updated_backfill = backfill.with_asset_backfill_data(
-        result.backfill_data, dynamic_partitions_store=instance
-    )
-    if result.backfill_data.is_complete():
-        # The asset backfill is complete when all runs to be requested have finished (success,
+        pipeline_and_execution_plan_cache: Dict[int, Tuple[ExternalJob, ExternalExecutionPlan]] = {}
+
+        for run_request in result.run_requests:
+            yield None
+            submit_run_request(
+                run_request=run_request,
+                asset_graph=asset_graph,
+                # create a new request context for each run in case the code location server
+                # is swapped out in the middle of the backfill
+                workspace=workspace_process_context.create_request_context(),
+                instance=instance,
+                pipeline_and_execution_plan_cache=pipeline_and_execution_plan_cache,
+            )
+
+    elif backfill.status == BulkActionStatus.CANCELING:
+        # The asset backfill is successfully canceled when all requested runs have finished (success,
         # failure, or cancellation). Since the AssetBackfillData object stores materialization states
         # per asset partition, the daemon continues to update the backfill data until all runs have
         # finished in order to display the final partition statuses in the UI.
-        updated_backfill = updated_backfill.with_status(BulkActionStatus.COMPLETED)
 
-    pipeline_and_execution_plan_cache: Dict[int, Tuple[ExternalJob, ExternalExecutionPlan]] = {}
+        # We check if the all runs are finished before updating the asset backfill data.
+        # If we did this the other way around, a race condition would occur where runs finish
+        # between the asset backfill data update and the backfill cancellation.
+        mark_backfill_as_canceled = asset_backfill_data.requested_runs_finished(
```

Comment:
Why do we need to check here for whether the requested runs have finished if we're also monitoring runs inside `get_canceling_asset_backfill_iteration_data`? It seems like we might be able to find out if we're done by checking whether materialized_subset | failed_and_downstream_subset == requested_subset ?

Reply:
good call. for some reason the fact that we had a requested subset as well slipped my mind